CMSgov/mint-app
# Easi 2338/data auditing

# EASI-2338

## Changes and Description

### Overview
This is a a fairly large PR that introduce an auditing paradigm to MINT. At a high level, it uses triggers to capture insert, delete, and update actions and right a record to the audit table. 

#### Changes that are captured
1. Who inserted or updated a table
    a. deleting a row does not capture who performed the action, as we don't know who performed the action
2. Specific configurable columns entries on insert / delete
   a. Otherwise, we capture a complete record of the initial entry every insert / or delete, which is not needed
3. All changes in a transaction besides columns that are configured to be ignored. (META data changes do not need a separate entry written)

- Implement an auditing schema
    - audit.change -> Store transactional changes
    - audit.table_config -> Store configurations that are used by the audit trigger to audit each table with configurable behavior
- Model_Plan Name History
   - return a string array defaulted in `DESC` order for a plans name history, with optional sort order param.
- Generic GQL endpoint to get a change by primary key, and table name
   - there is no immediate use case for this. It is likely the signature of this will change to not return a map, but instead a generated type, so you can return an array with field name, old value, and new value.
   
 ## Future development
To protect the scope of this change, it is necessary to note what will be developed in the future as needed.
1. Returning generic change history to the front end as an array instead of a map
2. SQL to query based on a search parameter object which can return collection based on shared secondary key (eg all task list sections and model plans in one transaction, as well as a date range)
3. Analysis based on audit changes with who is subscribed to a change.
4. Possibility of other indexes added to the change table as needed (ok primary key and foreign key)
5. Custom View which combines table_config and change.
   


## How to test this change
1. `scripts/dev up:backend:debug && scripts/dev db:seed`
2. Update a model plan ( or other object as you desire) ~ Note, you will need to update the id to an existing model plan ID ~
```gql
mutation modelPlanUpdate{
  
  updateModelPlan(id:"6e224030-09d5-46f7-ad04-4bb851b36eab",changes
      :{
            modelName: "Another Cool Name NEW",
            # modelCategory: DEMONSTRATION,
            # cmsCenters: [CMMI,OTHER] ,           
            # cmsOther: "Excellent Good Center",
            # cmmiGroups: [PATIENT_CARE_MODELS_GROUP, POLICY_AND_PROGRAMS_GROUP],
            status: OMB_ASRF_CLEARANCE
            archived: false

            
      }
      )
      
      {
        id

        modelName
        nameHistory

        archived
        status
        createdBy
        createdDts
        modifiedBy
        modifiedDts
    
  }  
  
}

```
3. Note that the above returns the name history of the model in `DESC` order.
6. If you want to query other changes, you can use this generic GQL endpoint, you can also query the database to see the structure
```gql
query auditChanges{
  auditChanges(tableName: "model_plan",primaryKey: "6e224030-09d5-46f7-ad04-4bb851b36eab")
   {
        id
        primaryKey
        foreignKey
        tableName
        action
        fields
        modifiedBy
        modifiedDts
  }
}
```


## PR Author Review Checklist



## PR Reviewer Guidelines
- It's best to pull the branch locally and test it, rather than just looking at the code online!
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
- Always make comments, even if it's minor, or if you don't understand why something was done.
